tests/std-checks/core/src/ptr.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Add contracts for functions from std::ptr.
+
+use std::ptr::NonNull;
+
+/// Create wrapper functions to standard library functions that contains their contract.
+pub mod contracts {
+    use super::*;
+    use kani::{ensures, implies, mem::*, modifies, requires};
+
+    #[ensures(implies!(ptr.is_null() => result.is_none()))]
+    #[ensures(implies!(!ptr.is_null() => result.is_some()))]
+    pub fn new<T>(ptr: *mut T) -> Option<NonNull<T>> {
+        NonNull::new(ptr)
+    }
+
+    /// # Safety:
+    /// When calling this method, you have to ensure that all the following is true:
+    ///
+    ///  - The pointer must be properly aligned.
+    ///  - It must be “dereferenceable” in the sense defined in the module documentation.
+    ///  - TODO: The pointer must point to an initialized instance of T.
+    ///
+    /// You must enforce Rust’s aliasing rules, since the returned lifetime 'a is arbitrarily chosen and does not
+    /// necessarily reflect the actual lifetime of the data. In particular, while this reference exists, the memory
+    /// the pointer points to must not get mutated (except inside UnsafeCell).
+    /// Taken from: <https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_ref>
+    #[requires(assert_valid_ptr(obj.as_ptr()))]
+    pub unsafe fn as_ref<'a, T>(obj: &NonNull<T>) -> &'a T {
+        obj.as_ref()
+    }
+
+    #[requires(!ptr.is_null())]
+    pub unsafe fn new_unchecked<T>(ptr: *mut T) -> NonNull<T> {
+        NonNull::<T>::new_unchecked(ptr)
+    }
+
+    /// Safety
+    ///
+    /// Behavior is undefined if any of the following conditions are violated:
+    ///   - dst must be valid for both reads and writes.
+    ///   - dst must be properly aligned.
+    ///   - dst must point to a properly initialized value of type T.
+    ///
+    /// Note that even if T has size 0, the pointer must be non-null and properly aligned.
+    #[requires(assert_valid_ptr(dst) && has_valid_value(dst))]
+    #[modifies(dst)]
+    pub unsafe fn replace<T>(dst: *mut T, src: T) -> T {
```

Comment:
A possible contract to this where `repr_eq` compares the byte representation of the input values.
```rust
#[ensures(repr_eq(old(unsafe { *dst }), ret)]
#[ensures(repr_eq(old(src), unsafe { *dst })]
```

Reply:
we don't have `old` implemented yet. so remove this for now.